KSES: Add test cases for "bogus comment" syntax not currently detected.

dmsnell opened the pull request:
In [[58418]](https://core.trac.wordpress.org/changeset/58418) a test was added without the `test_` prefix in its function
name, and because of that, it wasn't run in the test suite.
The prefix has been added to ensure that it runs.

In the original patch, due to a logical bug, a recursive loop to
transform the inside contents of the bogus comments was never run
more than once. This has been fixed.

This patch also includes one more case where `kses` wasn't
properly detecting the bogus comment state, and adds a test case
to cover this. It limits itself to some but not all constructions
of invalid markup declaration so that it doesn't conflict with
existing behaviors around those and other kinds of invalid comments.

See also #6840

mukeshpanchal27 commented on tests/phpunit/tests/kses.php:
```diff
@@ -1955,8 +1955,10 @@ public function wp_kses_preserves_html_comments( $html_comment, $expected_output
 		return array(
-			'Normative HTML comment'            => array( 'before<!-- this is a comment -->after', 'before<!-- this is a comment -->after' ),
-			'Closing tag with invalid tag name' => array( 'before<//not a tag>after', 'before<//not a tag>after' ),
+			'Normative HTML comment'                     => array( 'before<!-- this is a comment -->after', 'before<!-- this is a comment -->after' ),
+			'Closing tag with invalid tag name'          => array( 'before<//not a tag>after', 'before<//not a tag>after' ),
+			'Incorrectly opened comment (Markup declaration)' => array( 'before<!also not a tag>after', 'before<!also not a tag>after' ),
+			'Incorrectly opened comment (Question mark)' => array( 'before<?also not a tag>after', 'before<?also not a tag>after' ),
```
```suggestion
			'Normative HTML comment'                          => array( 'before<!-- this is a comment -->after', 'before<!-- this is a comment -->after' ),
			'Closing tag with invalid tag name'               => array( 'before<//not a tag>after', 'before<//not a tag>after' ),
			'Incorrectly opened comment (Markup declaration)' => array( 'before<!also not a tag>after', 'before<!also not a tag>after' ),
			'Incorrectly opened comment (Question mark)'      => array( 'before<?also not a tag>after', 'before<?also not a tag>after' ),
```

Arrow indentions.

dmsnell replied:
Thanks @mukeshpanchal27 for the suggestion, but I believe that the WPCS formatting rules prefer misalignment here. You can try changing them, and see it reject the PR.

dmsnell committed:
KSES: Fix tests and detection of HTML Bogus Comment spans.

In [58418] a test was added without the `test_` prefix in its function
name, and because of that, it wasn't run in the test suite.
The prefix has been added to ensure that it runs.

In the original patch, due to a logical bug, a recursive loop to
transform the inside contents of the bogus comments was never run
more than once. This has been fixed.

This patch also includes one more case where `kses` wasn't
properly detecting the bogus comment state, and adds a test case
to cover this. It limits itself to some but not all constructions
of invalid markup declaration so that it doesn't conflict with
existing behaviors around those and other kinds of invalid comments.

Props ellatrix, dmsnell.
See #61009.
Follow-up to [58418].

Co-authored-by: ellatrix <ellatrix@git.wordpress.org>

ellatrix commented on tests/phpunit/tests/kses.php:
```diff
@@ -1936,11 +1936,18 @@ public function filter_wp_kses_object_added_in_html_filter( $tags, $context ) {
+		$this->assertFalse(
+			current_user_can( 'unfiltered_html' ),
+			'Test needs to verify behavior without "unfiltered_html" capability.'
+		);
```
[wp_kses](https://developer.wordpress.org/reference/functions/wp_kses/) itself does not check user caps right? So this is not needed?

dmsnell replied:
I think that's right, but the test is somewhat useless if that capability is available. it was working properly already, but I didn't know for sure. maybe it's best to nix this part. thank you.

ellatrix commented on src/wp-includes/kses.php:
```diff
@@ -988,6 +988,9 @@ function wp_kses_split( $content, $allowed_html, $allowed_protocols ) {
 		|
 		</[^a-zA-Z][^>]*>  #  - Closing tags with invalid tag names.
+		|
+		<![^>]*>           #  - Invalid markup declaration nodes. Not all invalid nodes
```
Should we make the same [a-z] check here?

ellatrix commented on src/wp-includes/kses.php:
```diff
@@ -1114,22 +1117,30 @@ function wp_kses_split2( $content, $allowed_html, $allowed_protocols ) {
+		do {
+			$prev    = $content;
+			$content = wp_kses( $content, $allowed_html, $allowed_protocols );
+		} while ( $prev !== $content );
```
Is there a test case that fails in trunk and now passes in the PR as a result of this fixed loop?

dmsnell replied:
no. I think it's because we don't have a test case showing how one form of sanitization adds a new thing needing sanitization. given that this is copying the existing behavior I don't want to touch that, and I'm not even sure what the scenario would be where it would need to trigger this.

ellatrix replied:
For anyone else looking at this, this is just copying https://github.com/WordPress/wordpress-develop/pull/6833/files#diff-13e853d5729978a0e7936543d65ad1a3e120b994702c9559bf99c3ad857f5066R1153, but had to be rewritten to make the linter happy.

ellatrix reviewed (approved):
Great, clean and easy to understand change.

dmsnell commented:
Merged in [[58424]](https://core.trac.wordpress.org/changeset/58424)
https://github.com/wordpress/wordpress-develop/commit/b36ff6064a6795431eac65d5d1ade5a2e0f1ec05
